core/finality-grandpa/src/lib.rs

```diff
@@ -619,12 +645,26 @@ impl<Block: BlockT<Hash=H256>> GrandpaJustification<Block> {
 			}
 		};
 
-		let ancestry: HashMap<_, _> = justification.ancestry
+		match voter::validate_commit(
+			&justification.commit,
+			voters,
+			None,
+			chain,
```

Comment:
Doing this naively might expose a DoS vector. We might want to set up the commit-verification function so it lazily checks the ancestry proof only if the total weight of the commits is `> 2/3`.

Reply:
> Most votes will fall on a child or grandchild of that block in practice, but we haven't imported any descendents of the block yet.

I think I was wrong here; again because honest authorities don't vote past the block. But drawing on the ancestry in general is useful